### PR TITLE
Create schemas for users in granted databases

### PR DESCRIPTION
**Pull request by benjaminjb**
**Checklist:**
 - [ ] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [ ] Have you updated or added documentation for the change, as applicable?
 - [ ] Have you tested your changes on all related environments with successful results, as applicable?
   - [ ] Have you added automated tests?

**Type of Changes:**
 - [x] New feature
 - [ ] Bug fix
 - [ ] Documentation
 - [ ] Testing enhancement
 - [ ] Other

**What is the current behavior (link to any open issues here)?**

**What is the new behavior (if this is a feature change)?**
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

To help developers set up and connect quickly, the operator can now create schemas for `spec.users` without using an init SQL script.

This is a gated feature: to turn on set the FeatureGate `AutoCreateUserSchema=true`.

If turned on, a cluster can be annotated with
`postgres-operator.crunchydata.com/autoCreateUserSchema=true`.

If the feature is turned on and the cluster is annotated, PGO will create a schema named after the user in every database where that user has permissions.

(PG note: creating a schema with the same name as the user means that the PG `search_path` should not need to be updated, since `search_path` defaults to `"$user", public`.)

As with our usual pattern, the operator does not remove/delete PG objects (users, databases) that are removed from the spec.

NOTE: There are several schema names that would be dangerous to the cluster's operation; for instance, if you had pgbouncer enabled (which would create a `pgbouncer` schema) it would be dangerous to create a user named `pgbouncer` and use this feature to create a schema for that user.

**Other Information**:

Issues: [PGO-1333]
Fixes https://github.com/CrunchyData/postgres-operator/issues/3568

### CODE REVIEW

File: internal/postgres/users.go [hunk 1]
```diff
@@ -162,5 +170,76 @@ SELECT pg_catalog.format('GRANT ALL PRIVILEGES ON DATABASE %I TO %I',
...
+		_, _ = sql.WriteString(`SELECT UNNEST(ARRAY[:databases]);`)
+
+		stdout, stderr, err = exec.ExecInDatabasesFromQuery(ctx,
+			sql.String(),
+			strings.Join([]string{
+				// Quiet NOTICE messages from IF EXISTS statements.
+				// - https://www.postgresql.org/docs/current/runtime-config-client.html
+				`SET client_min_messages = WARNING;`,
+
+				"CREATE SCHEMA IF NOT EXISTS AUTHORIZATION :username;",
```
Comment (benjaminjb) [inline internal/postgres/users.go:231]: This creates a schema named :username, owned by :username.

Our only other schema that we lock down that I can see is `pgbouncer`, and we lock that with 

```sql
REVOKE ALL PRIVILEGES
    ON SCHEMA :"namespace" FROM PUBLIC, :"username";
 GRANT USAGE
    ON SCHEMA :"namespace" TO :"username";
```
--[cite](https://github.com/CrunchyData/postgres-operator/blob/master/internal/pgbouncer/postgres.go#L169-L172)

The difference being something like (rhino1 created with the `authorization` code, rhino2 granted/revoked as per other code):

```
postgres=# \dn+
                                       List of schemas
  Name  |       Owner       |           Access privileges            |      Description       
--------+-------------------+----------------------------------------+------------------------
 public | pg_database_owner | pg_database_owner=UC/pg_database_owner+| standard public schema
        |                   | =U/pg_database_owner                   | 
 rhino1 | rhino1            |                                        | 
 rhino2 | postgres          | postgres=UC/postgres                  +| 
        |                   | rhino2=U/postgres                      | 
(3 rows)
```

**Review by cbandy — CHANGES_REQUESTED** [review]

File: internal/postgres/users.go (hunk 1 above)
Comment (cbandy) [reply internal/postgres/users.go:231]: These permissions LGTM.

File: internal/postgres/users.go [hunk 2]
```diff
@@ -162,5 +170,76 @@ SELECT pg_catalog.format('GRANT ALL PRIVILEGES ON DATABASE %I TO %I',
...
+			}, "\n"),
+			map[string]string{
+				"databases": strings.Join(databases, ", "),
```
[the remark this answers is not in the document]
Comment (benjaminjb) [reply internal/postgres/users.go:234]: hmmm, and spec.users.databases has no requirements...

File: internal/postgres/users_test.go [hunk 3]
```diff
@@ -192,3 +197,52 @@ COMMIT;`))
...
+		) error {
+			calls++
+
+			b, err := io.ReadAll(stdin)
+			assert.NilError(t, err)
+
+			// The command strings will contain either of two possibilities, depending on the user called.
+			commands := strings.Join(command, ",")
+			re := regexp.MustCompile("--set=databases=\\[\"db1\"\\],--set=username=user-single-db|--set=databases=\\[\"db1\",\"db2\"\\],--set=username=user-multi-db")
+			assert.Assert(t, cmp.Regexp(re, commands))
```
Comment (cbandy) [inline internal/postgres/users_test.go:217]: 👍🏻 Interesting to see that the `postgres.Exec*` design makes it hard to test variables.

**Review by cbandy — APPROVED** [review]

**Pull request merged**